crypto/x509/x509_vfy.c

```diff
@@ -524,8 +524,8 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
             if (!verify_cb_cert(ctx, x, i, X509_V_ERR_PATH_LENGTH_EXCEEDED))
                 return 0;
         }
-        /* Increment path length if not self issued */
-        if (!(x->ex_flags & EXFLAG_SI))
+        /* Increment path length if not a self issued intermediate CA */
+        if (i == 0 || (x->ex_flags & EXFLAG_SI) == 0)
```

Comment:
Wouldn't it be better to fix the off-by-one calculation of path length?
e.g.
```
if (is_ca && (x->ex_flags & EXFLAG_SI) == 0)
plen++;
```

with
```
is_ca = X509_check_ca(x);
```
and
```
- (plen > (x->ex_pathlen + proxy_path_length + 1))) {
+ (plen > (x->ex_pathlen + proxy_path_length))) {
```
 Code easier to read with correct values used in calculations. And code able to abort if the leaf is violating the constraint, which would indicate something is wrong on the issuing CA.

Reply:
The code is not intended to fail to validate leaf-position CAs that have "CA:true" whose issuer has path length:0.  While the former cannot be used as intermediate CAs to issue  EE certs, they are valid leaf certificates, when used as such.  There is no "off by one", the patched code produces correct results with "plen" (at the time of the test) being the count of CAs in the path including the present CA, so a violation entails of `plen > ex_pathlen + 1` (ignoring the proxy stuff).

Reply:
If someone can point to a good reason to fail to validate chains leading to a leaf CA that **would** violate path length if it did occur as an intermediate, then perhaps we could reconsider the logic. But since that case would be caught by any correct validator implementation, it is not clear why we need to preclude such certificates being used as EE certificates. The `path:0` in the issuer is sufficient to effectively make the `CA:true` in the leaf irrelevant, and allows some EE certs that forgot to clear the CA bit to continue to work as EE certs and continue to fail if ever misused as CA certs.

Reply:
What I meant by off-by-one is that the implementation has `plen=1` for path length `0`, `plen=2` for path length `1` and so on, as per RFC5280 definition of path length (only authorities are applicable to path length, non-authorities are not part of path length).

(There is one more issue in the path length code a few lines higher up, the path length constrains are not validated for any certificate that is deemed self-issued. So for example path length constraints issued root authority are ignored completely. Maybe that should be a separate issue?)

Reply:
@vdukhovni Regarding Leaf certificate violating constraint not triggering an error; after a careful reading of RFC5820 I think the OpenSSL implementation is correct in that aspect. 
 Errors on path length constraint should be caught in "Prepare for Next Cert" part of the RFC algorithm, which **SHOULD NOT** executed for the leaf certificate. So the correct behaviour as defined by RFC is to not check leaf for path length constraint violations.

Reply:
Root CA certificates are self-signed not self-issued, and path length constraints cannot apply to them because they are never in the middle of the chain. There's no separate issue to open.

As for the "off by one", it makes no difference.  Sure, instead of *always* incrementing at depth 0, we could instead *never* increment at depth 0, and adjust the test accordingly, but that does not really change anything.